### PR TITLE
Fixed issue with the IS NOT for requesters, watchers and technicians for Ticket, Change and Problem

### PR DESCRIPTION
**Pull request by tomolimo**
Fixed issue with the IS NOT for requesters, watchers and technicians for Ticket, Change and Problem
Added tests for IS NOT

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #12455

### CODE REVIEW

File: src/Search.php [hunk 1]
```diff
@@ -4723,6 +4724,64 @@ public static function addWhere($link, $nott, $itemtype, $ID, $searchtype, $val,
...
+
+                    $join = '';
+                    $crit = '';
+                    $notstr = 'NOT';
+                    $equalstr = '=';
+                   // crit for equals and notequals
+                    if (in_array($searchtype, ['equals', 'notequals'])) {
+                        if ($val == 0) {
+                           // $val == 0 special value to tell that there is no choice (= empty) in the dropdown
+                            $equalstr = 'IS NOT NULL';
```
Comment (cedric-anne) [inline src/Search.php:4763]: Should this be adapted to search type ?
Please add a test case for this special case.
```suggestion
                            $equalstr = $searchtype === 'equals' ? 'IS NULL' : 'IS NOT NULL';
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]
Produced conditions seems OK. It is a really specific patch, as it is really complicated to produce global handling of this, I have nothing better to propose.

A functionnal test should be added to validate returned results (see `testDateBeforeOrNot()` for a way to simulate a search with criteria, and `testProblems()` for a way to check results).

File: src/Search.php (hunk 1 above)
Comment (tomolimo) [reply src/Search.php:4763]: Hello @cedric-anne 
In fact these lines will be executed only and only when there is one and only one 'NOT', so this case does not need to be coded.
When there is no 'NOT' or two 'NOT' (= 'AND NOT' and 'NOT EQUALS') then the legacy code will be executed.
Thank you,
Regards,
Tomolimo

**Comment by tomolimo** [comment]
> Produced conditions seems OK. It is a really specific patch, as it is really complicated to produce global handling of this, I have nothing better to propose.
>
> A functionnal test should be added to validate returned results (see testDateBeforeOrNot() for a way to simulate a search with criteria, and testProblems() for a way to check results).

Hello @cedric-anne 
I don't really understand as there are several tests that have been added, see: [tests/functionnal/Search.php](https://github.com/glpi-project/glpi/pull/13083/files#diff-a4d269cced9f626839c5dab4edc4bb14791244f4e9f67835ef4fa9558d825738), or I didn't understand the purpose of your proposal.

Thank you,
Regards,
Tomolimo

**Comment by cedric-anne** [comment]
@tomolimo 

Could you take a look on #13684? It should fix this problem globally, and not just on ITIL object actors.

**Comment by tomolimo** [comment]
Hello @cedric-anne 
I'll check that and I'll keep you informed
Thank you for the ping :)
Regards,
Tomolimo

**Comment by cedric-anne** [comment]
Replaced by #13684.

**Pull request closed without merging**